plugin/caffe/caffe_data_iter.cc

```diff
@@ -0,0 +1,264 @@
+
+    // Caffe seems to understand phase inside an "include {}" block
+    if (!param_.prototxt.has_phase()) {
+      if (param_.prototxt.include().size()) {
+        if (param_.prototxt.include(0).has_phase()) {
+          param_.prototxt.set_phase(param_.prototxt.include(0).phase());
+        }
+      }
+    }
+
+    std::string type = param_.prototxt.type();
+    caffe_data_layer_ = caffe::LayerRegistry<Dtype>::CreateLayer(param_.prototxt);
+    CHECK(caffe_data_layer_ != nullptr) << "Failed creating caffe data layer";
+    const size_t top_size = param_.prototxt.top_size();
+    if (top_size > 0) {
+      if(top_size > NR_SUPPORTED_TOP_ITEMS) {
+        LOG(WARNING) << "Too may \"top\" items, only two (one data, one label) are currently supported";
+      }
+      top_.reserve(top_size);
+      for (size_t x = 0; x < top_size; ++x) {
+        ::caffe::Blob<Dtype> *blob = new ::caffe::Blob<Dtype>();
+        cleanup_blobs_.push_back(std::unique_ptr<::caffe::Blob<Dtype>>(blob));
+        top_.push_back(blob);
+      }
+      caffe_data_layer_->SetUp(bottom_, top_);
+      const std::vector<int> &shape = top_[DATA]->shape();
+      const size_t shapeDimCount = shape.size();
+      if (shapeDimCount > 0) {
+        batch_size_ = shape[0];
+        if (shapeDimCount > 1) {
+          channels_ = shape[1];
+          if (shapeDimCount > 2) {
+            width_ = shape[2];
+            if (shapeDimCount > 3) {
+              height_ = shape[3];
+            }
+          }
+        }
+      }
+
+      if(top_size > DATA) {
+        if (param_.flat) {
+          batch_data_.shape_ = mshadow::Shape2(batch_size_, width_ * height_);
+        } else {
+          batch_data_.shape_ = mxnet::TShape(top_[DATA]->shape().begin(), top_[DATA]->shape().end());
+        }
+        batch_data_.stride_ = batch_data_.size(batch_data_.shape_.ndim() - 1);
+      }
+      out_.data.clear();
+      if(top_size > LABEL) {
+        batch_label_.shape_ = mxnet::TShape(top_[LABEL]->shape().begin(), top_[LABEL]->shape().end());
+        batch_label_.stride_ = 1;
+      }
+      out_.batch_size = batch_size_;
+    }
+  }
+
+  virtual void BeforeFirst(void) {
+    loc_ = 0;
+  }
+
+  virtual bool Next(void) {
+    // MxNet iterator is expected to return CPU-accessible memory
+    if(::caffe::Caffe::mode() != ::caffe::Caffe::CPU) {
+      ::caffe::Caffe::set_mode(::caffe::Caffe::CPU);
+      CHECK_EQ(::caffe::Caffe::mode(), ::caffe::Caffe::CPU);
+    }
+    caffe_data_layer_->Forward(bottom_, top_);
+    CHECK_GT(batch_size_, 0) << "batch size must be greater than zero";
+    CHECK_EQ(out_.batch_size, batch_size_) << "Internal Error: batch size mismatch";
+
+    if (loc_ + batch_size_ <= param_.num_examples) {
+      batch_data_.dptr_ = top_[DATA]->mutable_cpu_data();
+      batch_label_.dptr_ = top_[LABEL]->mutable_cpu_data();
+
+      out_.data.clear();
+      if (param_.flat) {
+        out_.data.push_back(batch_data_.FlatTo2D<cpu, Dtype>());
```

Comment:
I discussed this with Mu a couple of days ago and he suggested I just leave it in anyway.  I will remove the of check.